In [20]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

In [21]:
df = pd.read_csv('water_potability.csv')

# Question 1

Data Preprocessing


In [19]:
df.head()

ph    Hardness        Solids  Chloramines     Sulfate  Conductivity  \
0       NaN  204.890455  20791.318981     7.300212  368.516441    564.308654   
1  3.716080  129.422921  18630.057858     6.635246         NaN    592.885359   
2  8.099124  224.236259  19909.541732     9.275884         NaN    418.606213   
3  8.316766  214.373394  22018.417441     8.059332  356.886136    363.266516   
4  9.092223  181.101509  17978.986339     6.546600  310.135738    398.410813   

   Organic_carbon  Trihalomethanes  Turbidity  Potability  
0       10.379783        86.990970   2.963135           0  
1       15.180013        56.329076   4.500656           0  
2       16.868637        66.420093   3.055934           0  
3       18.436524       100.341674   4.628771           0  
4       11.558279        31.997993   4.075075           0

In [5]:
df.describe()

ph     Hardness        Solids  Chloramines      Sulfate  \
count  2785.000000  3276.000000   3276.000000  3276.000000  2495.000000   
mean      7.080795   196.369496  22014.092526     7.122277   333.775777   
std       1.594320    32.879761   8768.570828     1.583085    41.416840   
min       0.000000    47.432000    320.942611     0.352000   129.000000   
25%       6.093092   176.850538  15666.690297     6.127421   307.699498   
50%       7.036752   196.967627  20927.833607     7.130299   333.073546   
75%       8.062066   216.667456  27332.762127     8.114887   359.950170   
max      14.000000   323.124000  61227.196008    13.127000   481.030642   

       Conductivity  Organic_carbon  Trihalomethanes    Turbidity   Potability  
count   3276.000000     3276.000000      3114.000000  3276.000000  3276.000000  
mean     426.205111       14.284970        66.396293     3.966786     0.390110  
std       80.824064        3.308162        16.175008     0.780382     0.487849  
min      181.483754        2.200000         0.738000     1.450000     0.000000  
25%      365.734414       12.065801        55.844536     3.439711     0.000000  
50%      421.884968       14.218338        66.622485     3.955028     0.000000  
75%      481.792304       16.557652        77.337473     4.500320     1.000000  
max      753.342620       28.300000       124.000000     6.739000     1.000000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3276 entries, 0 to 3275
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ph               2785 non-null   float64
 1   Hardness         3276 non-null   float64
 2   Solids           3276 non-null   float64
 3   Chloramines      3276 non-null   float64
 4   Sulfate          2495 non-null   float64
 5   Conductivity     3276 non-null   float64
 6   Organic_carbon   3276 non-null   float64
 7   Trihalomethanes  3114 non-null   float64
 8   Turbidity        3276 non-null   float64
 9   Potability       3276 non-null   int64  
dtypes: float64(9), int64(1)
memory usage: 256.1 KB


In [7]:
df.isnull().sum()

ph                 491
Hardness             0
Solids               0
Chloramines          0
Sulfate            781
Conductivity         0
Organic_carbon       0
Trihalomethanes    162
Turbidity            0
Potability           0
dtype: int64

# Question 2

Outlier: An outlier is bservation in a dataset that significantly deviates from the majority of the data points.

Importance of Outliers:
1. Outliers can indicate errors in data collection or entry.
2. They can skew summary statistics and lead to incorrect conclusions.
3. Outliers can also contain valuable information or indicate interesting phenomena.

# Question 3

Three common ways of detecting outliers in the given dataset:

a. Z-Score Method: This method is used to identify outliers in a dataset by measuring the amount of stadard deviations the data point is away from the mean. It works by: 

    i. Calculating the mean(μ) and standard deviation(σ) of the dataset.
    
    ii. For each data point, the Z-Score is computed using the formula: Z = (X - μ) / σ, where X is the data point.
    
    iii. Normally, a threshold value which is normally set to 3 or -3, is used. Data points with Z-Scores that exceed the threshold are considered outliers.
    

b. Interquartile Range (IQR) Method: This method is also used to identify outliers and is not affected by extreme values. It works by:

    i. Calculating the first quartile (Q1) and third quartile (Q3) of the dataset.
        
    ii. Computing the IQR, which is the range between Q3 and Q1: IQR = Q3 - Q1.
    
    iii. Data points below Q1 - 1.5 * IQR or above Q3 + 1.5 * IQR are considered outliers.
        
c. Data Visualization (Box Plot): This method involves displaying the the distribution of the dataset, showing the median, quartiles, and potential outliers. The outliers are often represented as individual points beyond "whiskers" extending from the boxes. The data points that fall outside the whiskers are considered outliers.